# Welcome to A Gentle Introduction to Keras 

This course focuses on a specific sub-field of machine learning called **predictive modeling.**

Within predicitve modeling is a speciality or another sub-field called **deep learning.**

We will be crafting deep learning models with a library called Keras. 

>**Predictive modeling** is focused on developing models that make accurate predictions at the expense of explaining why predictions are made. 

You and I don't need to be able to write a binary classification model. We need to know how to use and interpret the results of the model. 

**Where does machine learning fit into data science?**

Data science is a much broader discipline. 

> Data Scientists take the raw data, analyse it, connect the dots and tell a story often via several visualizations. They usually have a broader range of skill-set and may not have too much depth into more than one or two. They are more on the creative side. Like an Artist. An Engineer, on the other hand, is someone who looks at the data as something they have to take in and churn out an output in some appropriate form in the most efficient way possible. The implementation details and other efficiency hacks are usually on the tip of their fingers. There can be a lot of overlap between the two but it is more like A Data Scientist is a Machine Learning Engineer but not the other way round. -- Ria Chakraborty, Data Scientist





# Step 1. Import our modules

Two important points here. Firstly, the **from** means we aren't importing the entire library, only a specific module. Secondly, notice we **are** imporing the entire numpy library. 

> If you get a message that states: WARNING (theano.configdefaults): g++ not detected, blah... blah. Run this in your Anaconda prompt. 

conda install mingw libpython



In [1]:
from keras.models import Sequential
from keras.layers import Dense
import numpy

C:\Users\Jergb\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


# Step 2.  Set our random seed


Run an algorithm on a dataset and you've built a great model. Can you produce the same model again given the same data?
You should be able to. It should be a requirement that is high on the list for your modeling project.

> We achieve reproducibility in applied machine learning by using the exact same code, data and sequence of random numbers.

Random numbers are created using a random number generator. It’s a simple program that generates a sequence of numbers that are random enough for most applications.

This math function is deterministic. If it uses the same starting point called a seed number, it will give the same sequence of random numbers.

Hold on... what's **deterministic** mean? 

> "a deterministic algorithm is an algorithm which, given a particular input, will always produce the same output, with the underlying machine always passing through the same sequence of states"

Let's apply an English translator to this: 

> The **only purpose of seeding** is to make sure that you get the **exact same result** when you run this code many times on the exact same data.

In [2]:
seed = 9
numpy.random.seed(seed)

# Step 3.  Import our data set


Let's import the object called read_csv. 

We define a variable called filename and put our data set in it. 

The last line does the work. It using the function called **read_csv** to put the contents of our data set into a variable called dataframe. 

In [3]:
from pandas import read_csv
filename = 'BBCN.csv'
dataframe = read_csv(filename)

# Step 4.  Split the Output Variables


The first thing we need to do is put our data in an array. 

> An array is a data structure that stores values of **same data type**. 

In Python, this is the main difference between arrays and lists. While python lists can contain values corresponding to different data types, arrays in python can only contain values corresponding to same data type.

In [4]:
array = dataframe.values

In [5]:
array.shape

(500, 12)

The code below is the trickest part of the exercise. Now, we are assinging X and y as output variables.

> That looks pretty easy but keep in mind that an array starts at 0. 

If you take a look at the shape of our dataframe (shape means the number of columns and rows) you can see we have 12 rows. 

On the X array below we saying... include all items in the array from 0 to 11. 

On the y array below we are saying... just use the column in the array mapped to the **11th row**. The **BikeBuyer** column. 

> Before we split X and Y out we are going to put them in an array. 




In [6]:
X = array[:,0:11] 
Y = array[:,11]

In [7]:
X.shape

(500, 11)

In [8]:
Y.shape

(500,)

In [9]:
dataframe.head()

MaritalStatus  Gender  YearlyIncome  TotalChildren  NumberChildrenAtHome  \
0              5       1           9.0              2                     0   
1              5       1           6.0              3                     3   
2              5       1           6.0              3                     3   
3              5       2           7.0              0                     0   
4              5       2           8.0              5                     5   

   EnglishEducation  HouseOwnerFlag  NumberCarsOwned  CommuteDistance  Region  \
0                 5               1                0                2       2   
1                 5               0                1                1       2   
2                 5               1                1                5       2   
3                 5               0                1               10       2   
4                 5               1                4                2       2   

   Age  BikeBuyer  
0    5          1  
1    4          1  
2    4          1  
3    5          1  
4    5          1

# Step 4.  Build the Model


We can piece it all together by adding each layer. 

> The first layer has 11 neurons and expects 11 input variables. 

The second hidden layer has 8 neurons.

The third hidden layer has 8 neurons. 

The output layer has 1 neuron to predict the class. 

How many hidden layers are in our model? 

In [10]:
model = Sequential()
model.add(Dense(12, input_dim=11, init='uniform', activation='relu'))
model.add(Dense(8, init='uniform', activation='relu'))
model.add(Dense(8, init='uniform', activation='relu'))
model.add(Dense(8, init='uniform', activation='relu'))
model.add(Dense(1, init='uniform', activation='sigmoid'))

C:\Users\Jergb\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(12, input_dim=11, activation="relu", kernel_initializer="uniform")`
  
C:\Users\Jergb\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(8, activation="relu", kernel_initializer="uniform")`
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\Jergb\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(8, activation="relu", kernel_initializer="uniform")`
  after removing the cwd from sys.path.
C:\Users\Jergb\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(8, activation="relu", kernel_initializer="uniform")`
  """
C:\Users\Jergb\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: UserWarning: Update your `Dense` call to t

# Step 5.  Compile the Model

A metric is a function that is used to judge the performance of your model. Metric functions are to be supplied in the  metrics parameter when a model is compiled.

>  Lastly, we set the cost (or loss) function to categorical_crossentropy. The (binary) cross-entropy is just the technical term for the **cost function** in logistic regression, and the categorical cross-entropy is its generalization for multi-class predictions via softmax

Binary learning models are models which just predict one of two outcomes: positive or negative. These models are very well suited to drive decisions, such as whether to administer a patient a certain drug or to include a lead in a targeted marketing campaign.

> Accuracy is perhaps the most intuitive performance measure. **It is simply the ratio of correctly predicted observations.**

Using accuracy is only good for symmetric data sets where the class distribution is 50/50 and the cost of false positives and false negatives are roughly the same. In our case our classes are balanced. 

Whenever you train a model with your data, you are actually producing some new values (predicted) for a specific feature. However, that specific feature already has some values which are real values in the dataset. 

> We know the the closer the predicted values to their corresponding real values, the better the model.

We are using cost function to measure **how close the predicted values are to their corresponding real values.**

So, for our model we choose binary_crossentropy. 

In [11]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
#model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

# Step 5.  Fit the Model

**Epoch:** A full pass over all of your training data.

For example, let's say you have 1213 observations. So an epoch concludes when it has finished a training pass over all 1213 of your observations.

> What you'd expect to see from running fit on your Keras model, is a decrease in loss over n number of epochs.

batch_size denotes the subset size of your training sample (e.g. 100 out of 1000) which is going to be used in order to train the network during its learning process. 

Each batch trains network in a successive order, taking into account the updated weights coming from the appliance of the previous batch. 

>Example: if you have 1000 training examples, and your batch size is 500, then it will take 2 iterations to complete 1 epoch.



In [12]:
model.fit(X, Y, nb_epoch=300, batch_size=30)

C:\Users\Jergb\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  """Entry point for launching an IPython kernel.


Epoch 1/300
500/500 [==============================] - 0s 787us/step - loss: 0.6931 - acc: 0.5220
Epoch 2/300
500/500 [==============================] - 0s 64us/step - loss: 0.6928 - acc: 0.5340
Epoch 3/300
500/500 [==============================] - 0s 60us/step - loss: 0.6926 - acc: 0.5340
Epoch 4/300
500/500 [==============================] - 0s 64us/step - loss: 0.6924 - acc: 0.5340
Epoch 5/300
500/500 [==============================] - 0s 64us/step - loss: 0.6922 - acc: 0.5340
Epoch 6/300
500/500 [==============================] - 0s 82us/step - loss: 0.6921 - acc: 0.5340
Epoch 7/300
500/500 [==============================] - 0s 58us/step - loss: 0.6919 - acc: 0.5340
Epoch 8/300
500/500 [==============================] - 0s 78us/step - loss: 0.6917 - acc: 0.5340
Epoch 9/300
500/500 [==============================] - 0s 60us/step - loss: 0.6914 - acc: 0.5340
Epoch 10/300
500/500 [==============================] - 0s 74us/step - loss: 0.6911 - acc: 0.5340
Epoch 11/300
500/500 [======

500/500 [==============================] - 0s 56us/step - loss: 0.5859 - acc: 0.6820
Epoch 85/300
500/500 [==============================] - 0s 66us/step - loss: 0.5831 - acc: 0.6740
Epoch 86/300
500/500 [==============================] - 0s 52us/step - loss: 0.5824 - acc: 0.6840
Epoch 87/300
500/500 [==============================] - 0s 52us/step - loss: 0.5809 - acc: 0.6760
Epoch 88/300
500/500 [==============================] - 0s 50us/step - loss: 0.5839 - acc: 0.6680
Epoch 89/300
500/500 [==============================] - 0s 56us/step - loss: 0.5821 - acc: 0.6700
Epoch 90/300
500/500 [==============================] - 0s 60us/step - loss: 0.5804 - acc: 0.6880
Epoch 91/300
500/500 [==============================] - 0s 58us/step - loss: 0.5800 - acc: 0.6880
Epoch 92/300
500/500 [==============================] - 0s 54us/step - loss: 0.5790 - acc: 0.6740
Epoch 93/300
500/500 [==============================] - 0s 72us/step - loss: 0.5746 - acc: 0.6920
Epoch 94/300
500/500 [===========

500/500 [==============================] - 0s 58us/step - loss: 0.5332 - acc: 0.7220
Epoch 168/300
500/500 [==============================] - 0s 62us/step - loss: 0.5298 - acc: 0.7320
Epoch 169/300
500/500 [==============================] - 0s 58us/step - loss: 0.5355 - acc: 0.7200
Epoch 170/300
500/500 [==============================] - 0s 50us/step - loss: 0.5347 - acc: 0.7280
Epoch 171/300
500/500 [==============================] - 0s 58us/step - loss: 0.5303 - acc: 0.7360
Epoch 172/300
500/500 [==============================] - 0s 54us/step - loss: 0.5294 - acc: 0.7260
Epoch 173/300
500/500 [==============================] - 0s 48us/step - loss: 0.5297 - acc: 0.7340
Epoch 174/300
500/500 [==============================] - 0s 50us/step - loss: 0.5315 - acc: 0.7220
Epoch 175/300
500/500 [==============================] - 0s 50us/step - loss: 0.5322 - acc: 0.7260
Epoch 176/300
500/500 [==============================] - 0s 50us/step - loss: 0.5308 - acc: 0.7280
Epoch 177/300
500/500 [=

500/500 [==============================] - 0s 50us/step - loss: 0.5121 - acc: 0.7360
Epoch 250/300
500/500 [==============================] - 0s 46us/step - loss: 0.5142 - acc: 0.7440
Epoch 251/300
500/500 [==============================] - 0s 52us/step - loss: 0.5141 - acc: 0.7320
Epoch 252/300
500/500 [==============================] - 0s 44us/step - loss: 0.5086 - acc: 0.7440
Epoch 253/300
500/500 [==============================] - 0s 50us/step - loss: 0.5092 - acc: 0.7400
Epoch 254/300
500/500 [==============================] - 0s 66us/step - loss: 0.5059 - acc: 0.7500
Epoch 255/300
500/500 [==============================] - 0s 46us/step - loss: 0.5076 - acc: 0.7440
Epoch 256/300
500/500 [==============================] - 0s 54us/step - loss: 0.5092 - acc: 0.7480
Epoch 257/300
500/500 [==============================] - 0s 60us/step - loss: 0.5060 - acc: 0.7420
Epoch 258/300
500/500 [==============================] - 0s 48us/step - loss: 0.5056 - acc: 0.7520
Epoch 259/300
500/500 [=

# Step 6.  Score the Model

In [13]:
scores = model.evaluate(X, Y)
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))
#78.4

500/500 [==============================] - 0s 148us/step
acc: 75.80%


500/500 [==============================] - 0s 140us/step  
acc: 73.20%

500/500 [==============================] - 0s 148us/step  
acc: 74.40%

500/500 [==============================] - 0s 136us/step  
acc: 73.60%

500/500 [==============================] - 0s 144us/step  
acc: 73.80%